### ISSUE TITLE
Links

### ISSUE THREAD

**Issue by coatk1** [coatk1/playground#96]

**Comment by coatk1** [comment]
https://www.anaconda.com/blog/get-python-package-download-statistics-with-condastats

**Comment by coatk1** [comment]
https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository

**Comment by coatk1** [comment]
https://hub.github.com/hub-api.1.html

**Comment by coatk1** [comment]
https://attack.mitre.org/
https://pythonforcybersecurity.com/courses/python3-for-infosec-professionals/
https://simpleprogrammer.com/python-programming-for-cybersecurity/
https://www.hackingloops.com/python-libraries-for-cybersecurity-development/

**Comment by coatk1** [comment]
https://www.tutorialspoint.com/python_digital_forensics/index.htm
https://www.tutorialspoint.com/python_forensics/index.htm

**Comment by coatk1** [comment]
https://realpython.com/python-logging/

**Comment by coatk1** [comment]
```
import logging
import datetime

# Config for logging to files.
logging.basicConfig(
    level=logging.INFO,
    filename="{}.log".format(datetime.date.today().strftime("%m_%d_%Y")),
    filemode="a",
    format="%(asctime)s - %(name)s - %(levelname)s: %(message)s",
    datefmt="I:%M:%S %p"
)

# Create handlers.
c_handler = logging.StreamHandler()
c_handler.setLevel(logging.INFO)

# Create formatters and add it to handlers.
c_format = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s: %(message)s", datefmt="I:%M:%S %p")
c_handler.setFormatter(c_format)

# Add handlers
logging.getLogger().addHandler(c_handler)

logger1 = logging.getLogger("main")
logger2 = logging.getLogger("secondary")

logger1.info("Test 1")
logger2.info("Test 2")
```

**Comment by coatk1** [comment]
https://learn.microsoft.com/en-us/azure/devops/repos/git/git-branching-guidance?view=azure-devops

**Comment by coatk1** [comment]
https://www.anaconda.com/blog/understanding-and-improving-condas-performance

**Comment by coatk1** [comment]
https://github.com/allenai/scispacy/issues/141

**Comment by coatk1** [comment]
```
## Source: https://github.com/allenai/scispacy/issues/141#issuecomment-518274586
## Author: https://github.com/phosseini
##   File: /mnt/Vancouver/apps/spacy/entity_options.py
##    Env: Python 3.7 venv:
##    Use:
##          import entity_options
##          from entity_options import get_entity_options
##          displacy.serve(doc, style="ent", options=get_entity_options(random_colors=True))
##    Ent: https://github.com/allenai/scispacy/issues/79#issuecomment-557766506 ## CRAFT entities

import random 

def get_entity_options(random_colors=False):
    """ generating color options for visualizing the named entities """

    def color_generator(number_of_colors):
        color = ["#"+''.join([random.choice('0123456789ABCDEF') for j in range(6)]) for i in range(number_of_colors)]
        return color

    entities = ["GGP", "SO", "TAXON", "CHEBI", "GO", "CL", "DNA", "CELL_TYPE", "CELL_LINE", "RNA", "PROTEIN", \
                "DISEASE", "CHEMICAL", "CANCER", "ORGAN", "TISSUE", "ORGANISM", "CELL", "AMINO_ACID", \
                "GENE_OR_GENE_PRODUCT", "SIMPLE_CHEMICAL", "ANATOMICAL_SYSTEM", "IMMATERIAL_ANATOMICAL_ENTITY", \
                "MULTI-TISSUE_STRUCTURE", "DEVELOPING_ANATOMICAL_STRUCTURE", "ORGANISM_SUBDIVISION", "CELLULAR_COMPONENT"]

    colors = {"ENT":"#E8DAEF"}

    if random_colors:
        color = color_generator(len(entities))
        for i in range(len(entities)):
            colors[entities[i]] = color[i]
    else:
        entities_cat_1 = {"GGP":"#F9E79F", "SO":"#F7DC6F", "TAXON":"#F4D03F", "CHEBI":"#FAD7A0", "GO":"#F8C471", "CL":"#F5B041"}
        entities_cat_2 = {"DNA":"#82E0AA", "CELL_TYPE":"#AED6F1", "CELL_LINE":"#E8DAEF", "RNA":"#82E0AA", "PROTEIN":"#82E0AA"}
        entities_cat_3 = {"DISEASE":"#D7BDE2", "CHEMICAL":"#D2B4DE"}
        entities_cat_4 = {"CANCER":"#ABEBC6", "ORGAN":"#82E0AA", "TISSUE":"#A9DFBF", "ORGANISM":"#A2D9CE", "CELL":"#76D7C4", \
                          "AMINO_ACID":"#85C1E9", "GENE_OR_GENE_PRODUCT":"#AED6F1", "SIMPLE_CHEMICAL":"#76D7C4", "ANATOMICAL_SYSTEM":"#82E0AA", \
                          "IMMATERIAL_ANATOMICAL_ENTITY":"#A2D9CE", "MULTI-TISSUE_STRUCTURE":"#85C1E9", "DEVELOPING_ANATOMICAL_STRUCTURE":"#A9DFBF", \
                          "ORGANISM_SUBDIVISION":"#58D68D", "CELLULAR_COMPONENT":"#7FB3D5"}

        entities_cats = [entities_cat_1, entities_cat_2, entities_cat_3, entities_cat_4]
        for item in entities_cats:
            colors = {**colors, **item}

    options = {"ents": entities, "colors": colors}
    # print(options)
    return options
```

**Comment by coatk1** [comment]
* https://www.altexsoft.com/blog/datascience/machine-learning-project-structure-stages-roles-and-tools/
* https://medium.com/analytics-vidhya/folder-structure-for-machine-learning-projects-a7e451a8caaa
* https://towardsdatascience.com/structuring-machine-learning-projects-be473775a1b6
* https://drivendata.github.io/cookiecutter-data-science/

**Comment by coatk1** [comment]
* https://stackoverflow.com/questions/34478398/import-local-function-from-a-module-housed-in-another-directory-with-relative-im

**Comment by coatk1** [comment]
https://xlsxwriter.readthedocs.io/example_pandas_autofilter.html

**Comment by coatk1** [comment]
https://medium.com/analytics-vidhya/boxing-mma-brazilian-jiu-jitsu-how-do-americans-fight-b6aba2692cb5

**Comment by coatk1** [comment]
Intro to NLP
---
* nltk
* gensim
* defaultdict
* itertools.chain.from_iterable()
* tf-idf

    * Term frequency = percentage share of the word compared to all tokens in the document
    * Inverse document frequency = logarithm of the total number of documents in a corpora divided by the number of documents containing the term

* spacy
* polyglot
* scikit-learn

    * CountVectorizer for preprocessing
    * Error with data in test that is not in training data. Add more training data or remove words from test that isnt in training data.

* Naive Bayes Classifier
* Sentiment Analysis?

**Comment by coatk1** [comment]
```
# Create the list of alphas: alphas
alphas = np.arange(0, 1, 0.1)

# Define train_and_predict()
def train_and_predict(alpha):
    # Instantiate the classifier: nb_classifier
    nb_classifier = MultinomialNB(alpha=alpha)
    # Fit to the training data
    nb_classifier.fit(tfidf_train, y_train)
    # Predict the labels: pred
    pred = nb_classifier.predict(tfidf_test)
    # Compute accuracy: score
    score = metrics.accuracy_score(y_test, pred)
    return score

# Iterate over the alphas and print the corresponding score
for alpha in alphas:
    print('Alpha: ', alpha)
    print('Score: ', train_and_predict(alpha))
    print()
```

**Comment by coatk1** [comment]
Sentiment Analysis
----
* textblob

**Comment by coatk1** [comment]
Intro to spaCy
----

**Comment by coatk1** [comment]
* https://spacy.io/
* https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/install.html
* https://huggingface.co/docs/transformers/index
* https://www.altexsoft.com/blog/datascience/how-to-organize-data-labeling-for-machine-learning-approaches-and-tools/
* https://confluence.atlassian.com/jirakb/how-to-edit-the-resolution-of-a-jira-issue-313467778.html
* https://mobidev.biz/blog/how-to-build-text-based-recommender-system-with-nlp
* https://sophiamyang.github.io/DS/basics/testing_for_data_scientists/testing_for_data_scientists.html
* https://towardsdatascience.com/data-science-workflows-notebook-to-production-26afc13442bb

**Comment by coatk1** [comment]
* https://spacy.io/
* https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/install.html
* https://huggingface.co/docs/transformers/index
* https://www.altexsoft.com/blog/datascience/how-to-organize-data-labeling-for-machine-learning-approaches-and-tools/
* https://confluence.atlassian.com/jirakb/how-to-edit-the-resolution-of-a-jira-issue-313467778.html
* https://mobidev.biz/blog/how-to-build-text-based-recommender-system-with-nlp
* https://sophiamyang.github.io/DS/basics/testing_for_data_scientists/testing_for_data_scientists.html
* https://towardsdatascience.com/data-science-workflows-notebook-to-production-26afc13442bb

**Comment by coatk1** [comment]
https://www.analyticsvidhya.com/blog/2021/06/13-important-requirements-of-a-laptop-for-data-science-tasks/

**Comment by coatk1** [comment]
https://stackoverflow.com/questions/64622841/using-conda-build-to-build-win32-and-win64-packages-in-1-go

**Comment by coatk1** [comment]
https://towardsdatascience.com/visualizing-survivors-tv-show-data-set-castaways-personalities-and-voting-teams-patterns-a56774dda2a6

**Comment by coatk1** [comment]
* https://plotly.com/python/
* https://plotly.com/python/reference/scatter/
* https://plotly.com/python/reference/layout/
* https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_connect-to-a-remote-jupyter-server

**Comment by coatk1** [comment]
https://www.alphr.com/computer-monitor-flickering/
https://scikit-learn.org/stable/user_guide.html
https://scikit-learn.org/stable/api/index.html
https://huggingface.co/


https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForSequenceClassification.forward.example-2
https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
https://towardsdatascience.com/clustering-sentence-embeddings-to-identify-intents-in-short-text-48d22d3bf02e
https://cohere.com/blog/text-embeddings
https://www.sbert.net/examples/applications/semantic-search/README.html


https://maartengr.github.io/BERTopic/index.html
https://huggingface.co/blog/bertopic
https://www.trychroma.com/
https://umap-learn.readthedocs.io/en/latest/parameters.html
https://hdbscan.readthedocs.io/en/latest/parameter_selection.html
https://scikit-learn.org/stable/modules/generated/sklearn.cluster.HDBSCAN.html


https://distinctipy.readthedocs.io/en/latest/usage.html
https://plotly.com/python-api-reference/index.html
https://plotly.com/python/
https://plotly.com/python/reference/scatter/
https://plotly.com/python/reference/layout/


https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_connect-to-a-remote-jupyter-server


https://huggingface.co/docs/transformers/en/tasks/token_classification
https://www.kaggle.com/code/thanish/bert-for-token-classification-ner-tutorial
https://discuss.huggingface.co/t/fine-tune-bert-with-two-classification-heads-next-to-each-other/9984
https://discuss.huggingface.co/t/how-do-i-change-the-classification-head-of-a-model/4720
https://stackoverflow.com/questions/62337074/adding-new-labels-to-an-already-trained-bert-model
https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L1176


https://pytorch.org/docs/stable/generated/torch.Tensor.unfold.html
https://pytorch.org/tutorials/beginner/basics/data_tutorial.html
https://pytorch.org/docs/stable/data.html
https://huggingface.co/docs/transformers/main_classes/trainer


https://www.geeksforgeeks.org/sweetviz-automated-exploratory-data-analysis-eda/
https://docs.profiling.ydata.ai/latest/

**Comment by coatk1** [comment]
https://cookiecutter-data-science.drivendata.org/v1/

**Comment by coatk1** [comment]
https://www.deeplearning.ai/short-courses/claude-code-a-highly-agentic-coding-assistant/
